design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,163 @@
+
+### Supporting multiple clouds
+
+When the user adds a virtual ref, they need to indicate the details of the platform to retrieve it. These include what cloud it is (so we know how to set up credentials), region, endpoint url, port and other settings. It looks a lot like our `S3Config` type (or an expanded version of that).
+
+### Supporting multiple credentials
+
+Credentials are a read time concern. When the user writes the virtual chunk, they don't necessarily know what credentials set will be used to read it.
+
+Credentials need to be specified later, when the chunks are going to be read. For example, by indicating in runtime configuration a set of credentials for each virtual chunk bucket/prefix/template.
+
+## Design
+
+### Explicit containers for virtual chunks
+
+* Each repo has a global list of `VirtualChunkContainers`
+
+```rust
+// this is pseudo-code
+struct VirtualChunkContainer {
+  name: String,  // something like model-output
+  platform: Platform,  // something like S3
+  url_template: String, // something like some-bucket/some-prefix/output/{}/{}.nc
+  default_template_arguments: Vec<String>,
+  cloud_client_config: Map<String, Any>,  // something like {endpoint_url: 8080}
+}
+```
+
+* Users can add to the list and edit members but not delete.
+
+### Global config
+
+* The list of chunk containers is stored in a new global `Config` object. There is a single configuration per repo, and all reads and writes are done reading the latest configuration. In the initial implementation, the "history" of configuration is present only for reference and user inspection by hand.
+
+TBD: how to store this information. One approach would be using the new PUT if-match S3 functionality, but this would be the first place where we overwrite objects.
+
+```rust
+// this is pseudo-code
+struct Config {
+  inline_chunk_threshold_bytes: u16,
+  virtual_chunk_containers: Vec<VirtualChunkContainer>,
+}
+```
+
+* The `Config` object can act as a configuration default when users don't override. But different fields of configuration may have different behaviors.
+
+### New virtual chunk ref
+
+We change from
+
+```rust
+pub struct VirtualChunkRef {
+    pub location: VirtualChunkLocation, // this is basically a String for the URL today
+    pub offset: ChunkOffset,
+    pub length: ChunkLength,
+}
+```
+
+To
+
+```rust
+pub struct VirtualChunkRef {
+    pub container_index: u16,
+    pub container_template_args: Vec<Option<String>>,
+    pub offset: ChunkOffset,
+    pub length: ChunkLength,
+    pub last_modified: Option<u32>,
+}
+```
+
+* The `last_modified` field contains an optional unix epoch in seconds. If set for a chunk, retrieval will fail if the file was modified after this time. We use this instead of an ETag or checksum to save space on the manifests. `u32` should be enough until year ~ 2100.
```

Comment:
Appreciating that using an optional unix epoch saves on space, what do you expect the impact of chosing unix epoch over ETag or checksum to be on read or write time?

Reply:
I don't expect any performance impact on reads. The object store will retrieve the object only if the "didn't change" conditions are met without any observable latency difference. On writes, it's a different story, we need to discuss with @TomNicholas if the new proposal is something we can implement in VirtualiZarr and what the impact could be.

In any case, I'm making the checksum optional at the chunk level.